enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,281 @@
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+- This phase of the enhancement does not address modifying alerting rule
+  expressions or thresholds.  This may be addressed by a later enhancement,
+  possibly with something similar to the [time series as alert thresholds][2]
+  pattern.
+  - Users still have the option of creating a new `AlertingRule` resource based
+    on an existing platform rule, making their modifications there, and
+    silencing the original alert.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with two new resources types:
+`AlertingRule` and `AlertRelabelConfig`, both in the `monitoring.openshift.io`
+group.
+
+The new `AlertingRule` type will allow additional user-defined alerting rules,
+but without the ability to create recording rules provided by the existing
+`PrometheusRule` resource type. Example:
+
+```yaml
+apiVersion: v1alpha1
+kind: alertingrule.monitoring.openshift.io
+metadata:
+  name: custom-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: CustomGroup
+    rules:
+    - alert: CustomAlert
+      expr: "vector(1)"
+      for: 15m
+      labels:
+        custom: "true"
+```
+
+A new controller in the cluster-monitoring-operator will generate
+`PrometheusRule` resources based on the `AlertingRule` instances in the
+`openshift-monitoring` namespace.  These will be consumed by the platform
+prometheus-operator instance like any other `PrometheusRule`.  All user-defined
+alerts will be labeled identifying them as such -- likely using the existing
+`openshift_io_alert_source` label.
+
+The new `AlertRelabelConfig` type will allow relabeling operations on alerts.
+All `AlertRelabelConfig` instances will be collected in order to build a single
+secret with additional alert relabel configs used by the prometheus-operator.
+Example of changing an alert severity from `none` to `critical`:
+
+```yaml
+apiVersion: v1alpha1
+kind: alertrelabelconfig.monitoring.openshift.io
+metadata:
+  name: severity-change
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: SeverityChanges
+    configs:
+    - source_labels: [alertname, severity]
+      regex: "Watchdog;none"
+      target_label: severity
+      replacement: critical
+      action: replace
+```
+
+A new controller in the cluster-monitoring-operator will combine all
+`AlertRelabelConfig` instances in the `openshift-monitoring` namespace into a
+single secret, concatenating each group in lexigraphical order by name.  This
+secret will be referenced in the `additionalAlertRelabelConfigs` parameter for
+the platform Prometheus instance.
+
+The `AlertRelabelConfig` controller may optionally disable certain actions like
+`drop` to prevent users disabling critical platform alerts.  Silences already
+exist as a method of ignoring unwanted alerts.
```

Comment:
The alerts would still be evaluated by Prometheus and dropping an alert means that Prometheus doesn't send it to Alertmanager, I don't think that we should forbid `drop` actions.

Reply:
I think it will also allow a cluster admin to override an existing alert definition: they can deploy a modified version of a product-defined alerting rule and add an relabeling config to drop the product-defined alert.

Reply:
I don't have a strong opinion. Users can of course also silence the alerts instead of dropping them, but maybe that is problematic in some cases.